File: benchmarking/bench-cluster/php/BenchCore.php
<?php

use Statsig\Statsig;
use Statsig\StatsigUser;
use Statsig\StatsigOptions;

class BenchCore {
    static $SCRAPI_URL = "http://scrapi:8000";
    static $ITER_LITE = 1000;
    static $ITER_HEAVY = 10_000;
    static $SDK_TYPE = "statsig-server-core-php";

    public static function run() {
        $sdkVersion = self::getSdkVersion();
        echo "Statsig PHP Core (v" . $sdkVersion . ")\n";
        echo "--------------------------------\n";

        $specNames = self::loadSpecNames();

        $options = new StatsigOptions(
            /* specs_url */ self::$SCRAPI_URL . "/v2/download_config_specs",
            /* log_event_url */ self::$SCRAPI_URL . "/v1/log_event",
            /* specs_adapter */ null,
            /* event_logging_adapter */ null,
            /* environment */ null,
            /* event_logging_flush_interval_ms */ null,
            /* event_logging_max_queue_size */ null,
            /* specs_sync_interval_ms */ null,
            /* output_log_level */ null,
            /* disable_country_lookup */ null,
            /* wait_for_country_lookup_init */ null,
            /* wait_for_user_agent_init */ null,
            /* disable_all_logging */ null,
        );

        $statsig = new Statsig("secret-PHP_CORE", $options);
        $statsig->initialize();

        $globalUser = new StatsigUser("global_user");
        $results = [];

        foreach ($specNames['feature_gates'] as $gate) {
            $results[] = self::benchmark("check_gate", $gate, self::$ITER_HEAVY, function() use ($statsig, $gate) {
                $result = $statsig->checkGate(self::createUser(), $gate);
                if ($gate === "test_public" && $result === false) {
                    throw new Exception("test_public is false");
                }
            });

            $results[] = self::benchmark("check_gate_global_user", $gate, self::$ITER_HEAVY, function() use ($statsig, $gate, $globalUser)  {
                $statsig->checkGate($globalUser, $gate);
            });

            $results[] = self::benchmark("get_feature_gate", $gate, self::$ITER_HEAVY, function() use ($statsig, $gate) {
                $statsig->getFeatureGate(self::createUser(), $gate);
            });

            $results[] = self::benchmark("get_feature_gate_global_user", $gate, self::$ITER_HEAVY, function() use ($statsig, $gate, $globalUser) {
                $statsig->getFeatureGate($globalUser, $gate);
            });
        }

        foreach ($specNames['dynamic_configs'] as $config) {
            $results[] = self::benchmark("get_dynamic_config", $config, self::$ITER_HEAVY, function() use ($statsig, $config) {
                $statsig->getDynamicConfig(self::createUser(), $config);
            });

            $results[] = self::benchmark("get_dynamic_config_global_user", $config, self::$ITER_HEAVY, function() use ($statsig, $config, $globalUser) {
                $statsig->getDynamicConfig($globalUser, $config);
            });
        }

        foreach ($specNames['experiments'] as $experiment) {
            $results[] = self::benchmark("get_experiment", $experiment, self::$ITER_HEAVY, function() use ($statsig, $experiment) {
                $statsig->getExperiment(self::createUser(), $experiment);
            });

            $results[] = self::benchmark("get_experiment_global_user", $experiment, self::$ITER_HEAVY, function() use ($statsig, $experiment, $globalUser) {
                $statsig->getExperiment($globalUser, $experiment);
            });
        }

        foreach ($specNames['layers'] as $layer) {
            $results[] = self::benchmark("get_layer", $layer, self::$ITER_HEAVY, function() use ($statsig, $layer) {
                $statsig->getLayer(self::createUser(), $layer);
            });

            $results[] = self::benchmark("get_layer_global_user", $layer, self::$ITER_HEAVY, function() use ($statsig, $layer, $globalUser) {
                $statsig->getLayer($globalUser, $layer);
            });
        }

        $results[] = self::benchmark("get_client_initialize_response", "n/a", self::$ITER_LITE, function() use ($statsig) {
            $statsig->getClientInitializeResponse(self::createUser());
        });

        $results[] = self::benchmark("get_client_initialize_response_global_user", "n/a", self::$ITER_LITE, function() use ($statsig, $globalUser) {
            $statsig->getClientInitializeResponse($globalUser);
        });

        $statsig->shutdown();

        self::writeResults($results, $sdkVersion);
    }

    function loadSpecNames() {
        $path = "/shared-volume/spec_names.json";
        for ($i = 0; $i < 10; $i++) {
            if (file_exists($path)) { break; }
            echo "Waiting for spec_names.json to be created...\n";
            sleep(1);
        }
        echo "Loading spec_names.json...\n";
        $content = file_get_contents($path);
        $json = json_decode($content, true);
        echo "JSON: " . count($json['feature_gates']) . "\n";
        return $json;
    }

    function createUser() {
        return new StatsigUser(
            /* user_id */ 'user_' . rand(0, 1000000),
            /* custom_ids */ [],
            /* email */ "test@test.com",
            /* ip */ "127.0.0.1",
            /* user_agent */ "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36",
            /* country */ "US",
            /* locale */ "en-US",
            /* app_version */ "1.0.0",
            /* custom */ [
                "isAdmin" => false,
            ],
            /* private_attributes */ [
                "isPaid" => "nah",
            ],
        );
    }

    function benchmark(string $benchmarkName, string $specName, int $iterations, callable $func) {
        $durations = [];
        for ($i = 0; $i < $iterations; $i++) {
            $startMs = microtime(true) * 1000;
            $func();
            $endMs = microtime(true) * 1000;
            $durations[] = $endMs - $startMs;
        }

        sort($durations);
        $p99 = $durations[intval($iterations * 0.99)];
        $max = $durations[count($durations) - 1];
        $min = $durations[0];
        $median = $durations[intval($iterations / 2)];
        $avg = array_sum($durations) / $iterations;

        assert($min <= $max);

        $sdkVersion = self::getSdkVersion();

        $result = [
            "benchmarkName" => $benchmarkName,
            "p99" => $p99,
            "max" => $max,
            "min" => $min,
            "median" => $median,
            "avg" => $avg,
            "specName" => $specName,
            "sdkType" => self::$SDK_TYPE,
            "sdkVersion" => $sdkVersion,
        ];

        echo str_pad($benchmarkName, 30) . " p99(" . number_format($p99, 4) . "ms) max(" . number_format($max, 4) . "ms) " . $specName . "\n";

        return $result;
    }

    function getSdkVersion() {
        $installedPackages = json_decode(file_get_contents(__DIR__ . '/vendor/composer/installed.json'), true);
        $statsigCoreVersion = null;

        foreach ($installedPackages['packages'] as $package) {
            if ($package['name'] === 'statsig/statsig-php-core') {
                $statsigCoreVersion = $package['version'];
                break;
            }
        }

        return $statsigCoreVersion;
    }

    function writeResults(array $results, string $sdkVersion) {
        $path = "/shared-volume/" . self::$SDK_TYPE . "-" . $sdkVersion . "-results.json";
        file_put_contents($path, json_encode([
            "sdkType" => self::$SDK_TYPE,
            "sdkVersion" => $sdkVersion,
            "results" => $results,
        ], JSON_PRETTY_PRINT));
    }
}

File: benchmarking/bench-cluster/python/bench-core.py
import asyncio
import random
import string
import time
from typing import List, Dict, Any, Callable, Optional
import os
from statsig_python_core import (
    Statsig,
    StatsigUser,
    StatsigOptions,
)
from importlib.metadata import version
import json

sdk_type = "statsig-server-core-python"
sdk_version = version("statsig-python-core")

SCRAPI_URL = "http://scrapi:8000"
ITER_ULTRA_LITE = 100
ITER_LITE = 1000
ITER_HEAVY = 10_000


class BenchmarkResult:
    def __init__(
        self,
        benchmark_name: str,
        p99: float,
        max: float,
        min: float,
        median: float,
        avg: float,
        spec_name: str,
        sdk_type: str,
        sdk_version: str,
    ):
        self.benchmark_name = benchmark_name
        self.p99 = p99
        self.max = max
        self.min = min
        self.median = median
        self.avg = avg
        self.spec_name = spec_name
        self.sdk_type = sdk_type
        self.sdk_version = sdk_version

    def to_dict(self) -> Dict[str, Any]:
        return {
            "benchmarkName": self.benchmark_name,
            "p99": self.p99,
            "max": self.max,
            "min": self.min,
            "median": self.median,
            "avg": self.avg,
            "specName": self.spec_name,
            "sdkType": self.sdk_type,
            "sdkVersion": self.sdk_version,
        }


def setup():
    # Wait for spec_names.json to be available
    spec_name_path = "/shared-volume/spec_names.json"
    for i in range(10):
        if os.path.exists(spec_name_path):
            break
        time.sleep(1)

    with open(spec_name_path, "r") as f:
        return json.load(f)


def create_user() -> StatsigUser:
    user_id = "".join(random.choices(string.ascii_lowercase + string.digits, k=13))

    return StatsigUser(
        user_id=user_id,
        email="user@example.com",
        ip="127.0.0.1",
        locale="en-US",
        app_version="1.0.0",
        country="US",
        user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36",
        custom={"isAdmin": False},
        private_attributes={"isPaid": "nah"},
    )


def assert_cond(condition: bool, message: str):
    if not condition:
        raise Exception(message)


async def benchmark(
    bench_name: str,
    spec_name: str,
    iterations: int,
    func: Callable,
    results: List[BenchmarkResult],
    cleanup: Optional[Callable] = None,
):
    durations = []

    for _ in range(iterations):
        start = time.perf_counter()
        result = func()
        end = time.perf_counter()
        durations.append((end - start) * 1000)
        if cleanup:
            cleanup(result)

    # Calculate p99
    durations.sort()
    p99_index = int(iterations * 0.99)

    result = BenchmarkResult(
        benchmark_name=bench_name,
        p99=durations[p99_index],
        max=durations[-1],
        min=durations[0],
        median=durations[len(durations) // 2],
        avg=sum(durations) / len(durations),
        spec_name=spec_name,
        sdk_type=sdk_type,
        sdk_version=sdk_version,
    )

    results.append(result)

    print(
        f"{bench_name:<30} "
        f"p99({result.p99:.4f}ms){'':<15} "
        f"max({result.max:.4f}ms){'':<15} "
        f"{spec_name}"
    )

    await asyncio.sleep(0.001)  # 1ms delay


async def main():
    # ------------------------------------------------------------------------ [ Setup ]
    spec_names = setup()

    options = StatsigOptions(
        specs_url=f"{SCRAPI_URL}/v2/download_config_specs",
        log_event_url=f"{SCRAPI_URL}/v1/log_event",
    )

    statsig = Statsig("secret-PYTHON_CORE", options)
    statsig.initialize().wait()

    results = []

    # ------------------------------------------------------------------------ [ Benchmark ]

    print(f"Statsig Python Core (v{sdk_version})")
    print("--------------------------------")

    global_user = StatsigUser(user_id=f"global_user")

    def init_new_statsig(options: StatsigOptions, sdk_key: str):
        inst = Statsig(sdk_key, options)
        inst.initialize().wait()
        return inst

    await benchmark(
        "initialize",
        "proto",
        ITER_ULTRA_LITE,
        lambda: init_new_statsig(options, "secret-PYTHON_CORE::BC_USE_PROTO"),
        results,
        cleanup=lambda inst: inst.shutdown().wait(),
    )

    await benchmark(
        "initialize",
        "json",
        ITER_ULTRA_LITE,
        lambda: init_new_statsig(options, "secret-PYTHON_CORE::BC_USE_JSON"),
        results,
        cleanup=lambda inst: inst.shutdown().wait(),
    )

    # ------------------------------------------------------------------------ [ Benchmark Feature Gates ]

    for gate in spec_names["feature_gates"]:
        await benchmark(
            "check_gate",
            gate,
            ITER_HEAVY,
            lambda g=gate: statsig.check_gate(create_user(), g),
            results,
        )

        await benchmark(
            "check_gate_global_user",
            gate,
            ITER_HEAVY,
            lambda g=gate: statsig.check_gate(global_user, g),
            results,
        )

        await benchmark(
            "get_feature_gate",
            gate,
            ITER_HEAVY,
            lambda g=gate: statsig.get_feature_gate(create_user(), g),
            results,
        )

        await benchmark(
            "get_feature_gate_global_user",
            gate,
            ITER_HEAVY,
            lambda g=gate: statsig.get_feature_gate(global_user, g),
            results,
        )

    # ------------------------------------------------------------------------ [ Benchmark Dynamic Configs ]

    for config in spec_names["dynamic_configs"]:
        await benchmark(
            "get_dynamic_config",
            config,
            ITER_HEAVY,
            lambda c=config: statsig.get_dynamic_config(create_user(), c),
            results,
        )

        await benchmark(
            "get_dynamic_config_global_user",
            config,
            ITER_HEAVY,
            lambda c=config: statsig.get_dynamic_config(global_user, c),
            results,
        )

    config = statsig.get_dynamic_config(global_user, "operating_system_config")
    await benchmark(
        "get_dynamic_config_params",
        "string",
        ITER_HEAVY,
        lambda: assert_cond(
            config.get_string("str", "err") != "err", "string value is err"
        ),
        results,
    )
    await benchmark(
        "get_dynamic_config_params",
        "number",
        ITER_HEAVY,
        lambda: assert_cond(config.get_integer("num", 0) != 0, "number value is 0"),
        results,
    )
    await benchmark(
        "get_dynamic_config_params",
        "object",
        ITER_HEAVY,
        lambda: assert_cond(
            (
                config.get_object_json("obj", "") != ""
                if hasattr(config, "get_object_json")
                else config.get_object("obj", {}) != {}
            ),
            "object value is empty",
        ),
        results,
    )
    await benchmark(
        "get_dynamic_config_params",
        "array",
        ITER_HEAVY,
        lambda: assert_cond(
            (
                config.get_array_json("arr", "") != ""
                if hasattr(config, "get_array_json")
                else config.get_array("arr", []) != []
            ),
            "array value is empty",
        ),
        results,
    )

    # ------------------------------------------------------------------------ [ Benchmark Experiments ]

    for experiment in spec_names["experiments"]:
        await benchmark(
            "get_experiment",
            experiment,
            ITER_HEAVY,
            lambda e=experiment: statsig.get_experiment(create_user(), e),
            results,
        )

        await benchmark(
            "get_experiment_global_user",
            experiment,
            ITER_HEAVY,
            lambda e=experiment: statsig.get_experiment(global_user, e),
            results,
        )

    experiment = statsig.get_experiment(global_user, "experiment_with_many_params")
    await benchmark(
        "get_experiment_params",
        "string",
        ITER_HEAVY,
        lambda: assert_cond(
            experiment.get_string("a_string", "err") != "err", "string value is err"
        ),
        results,
    )
    await benchmark(
        "get_experiment_params",
        "object",
        ITER_HEAVY,
        lambda: assert_cond(
            (
                experiment.get_object_json("an_object", "") != ""
                if hasattr(experiment, "get_object_json")
                else experiment.get_object("an_object", {}) != {}
            ),
            "object value is empty",
        ),
        results,
    )
    await benchmark(
        "get_experiment_params",
        "array",
        ITER_HEAVY,
        lambda: assert_cond(
            (
                experiment.get_array_json("an_array", "") != ""
                if hasattr(experiment, "get_array_json")
                else experiment.get_array("an_array", []) != []
            ),
            "array value is empty",
        ),
        results,
    )

    # ------------------------------------------------------------------------ [ Benchmark Layers ]

    for layer in spec_names["layers"]:
        await benchmark(
            "get_layer",
            layer,
            ITER_HEAVY,
            lambda l=layer: statsig.get_layer(create_user(), l),
            results,
        )

        await benchmark(
            "get_layer_global_user",
            layer,
            ITER_HEAVY,
            lambda l=layer: statsig.get_layer(global_user, l),
            results,
        )

    layer = statsig.get_layer(global_user, "layer_with_many_params")
    await benchmark(
        "get_layer_params",
        "string",
        ITER_HEAVY,
        lambda: assert_cond(
            layer.get_string("a_string", "err") != "err", "string value is err"
        ),
        results,
    )
    await benchmark(
        "get_layer_params",
        "object",
        ITER_HEAVY,
        lambda: assert_cond(
            (
                layer.get_object_json("an_object", "") != ""
                if hasattr(layer, "get_object_json")
                else layer.get_object("an_object", {}) != {}
            ),
            "object value is empty",
        ),
        results,
    )
    await benchmark(
        "get_layer_params",
        "array",
        ITER_HEAVY,
        lambda: assert_cond(
            (
                layer.get_array_json("an_array", "") != ""
                if hasattr(layer, "get_array_json")
                else layer.get_array("an_array", []) != []
            ),
            "array value is empty",
        ),
        results,
    )

    # ------------------------------------------------------------------------ [ Benchmark GCIR ]

    await benchmark(
        "get_client_initialize_response",
        "n/a",
        ITER_LITE,
        lambda: statsig.get_client_initialize_response(create_user()),
        results,
    )

    await benchmark(
        "get_client_initialize_response_global_user",
        "n/a",
        ITER_LITE,
        lambda: statsig.get_client_initialize_response(global_user),
        results,
    )

    # ------------------------------------------------------------------------ [ Teardown ]

    statsig.shutdown().wait()

    results_file = f"/shared-volume/{sdk_type}-{sdk_version}-results.json"
    with open(results_file, "w") as f:
        json.dump(
            {
                "sdkType": sdk_type,
                "sdkVersion": sdk_version,
                "results": [result.to_dict() for result in results],
            },
            f,
            indent=2,
        )


if __name__ == "__main__":
    asyncio.run(main())
